File: validation/dsh/plugin/src/output-constraints.ts
/** 边界类型和展示；计数、约束规则与通过判定全部由 Python 负责。 */
import type { JsonSchema } from './contracts.js'

export interface OutputConstraints {
  maxLength: number
  unit: 'unicode-code-points' | 'utf8-bytes'
  countWhitespace: boolean
}
export interface FormatValidation {
  schema_version: 'output-length-check-v1'
  status: 'not-requested' | 'not-evaluated' | 'passed' | 'failed'
  constraints: OutputConstraints | null
  passed: boolean | null
  actual_length: number | null
  output_sha256: string | null
}
function object(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === 'object' && !Array.isArray(value)
}
export function decodeOutputConstraints(raw: unknown): OutputConstraints {
  if (!object(raw) || Object.keys(raw).sort().join(',') !== 'countWhitespace,maxLength,unit'
    || typeof raw.maxLength !== 'number' || !Number.isFinite(raw.maxLength)
    || (raw.unit !== 'unicode-code-points' && raw.unit !== 'utf8-bytes')
    || typeof raw.countWhitespace !== 'boolean') throw new Error('invalid outputConstraints fields')
  return { maxLength: raw.maxLength, unit: raw.unit, countWhitespace: raw.countWhitespace }
}
export function decodeFormatValidation(raw: unknown): FormatValidation {
  if (!object(raw) || raw.schema_version !== 'output-length-check-v1'
    || !['not-requested', 'not-evaluated', 'passed', 'failed'].includes(String(raw.status))
    || (raw.passed !== null && typeof raw.passed !== 'boolean')
    || (raw.actual_length !== null && (typeof raw.actual_length !== 'number'
      || !Number.isSafeInteger(raw.actual_length) || raw.actual_length < 0))
    || (raw.output_sha256 !== null && (typeof raw.output_sha256 !== 'string'
      || !/^[a-f0-9]{64}$/.test(raw.output_sha256)))) throw new Error('invalid format validation result')
  return { schema_version: 'output-length-check-v1', status: raw.status as FormatValidation['status'],
    constraints: raw.constraints === null ? null : decodeOutputConstraints(raw.constraints),
    passed: raw.passed as boolean | null, actual_length: raw.actual_length as number | null,
    output_sha256: raw.output_sha256 as string | null }
}
export function formatValidationSummary(value: unknown): string {
  if (value === undefined) return '未提供'
  const check = decodeFormatValidation(value)
  const labels = { 'not-requested': '未设置', 'not-evaluated': '未检查', passed: '通过', failed: '未通过' }
  if (check.actual_length === null || check.constraints === null) return labels[check.status]
  return `${labels[check.status]}，${check.actual_length}/${check.constraints.maxLength} `
    + `${check.constraints.unit === 'utf8-bytes' ? 'UTF-8 字节' : 'Unicode 码点'}`
    + `（${check.constraints.countWhitespace ? '计入' : '不计'}空白）`
}
export const OUTPUT_CONSTRAINTS_SCHEMA: JsonSchema = {
  type: 'object', additionalProperties: false, required: ['maxLength', 'unit', 'countWhitespace'],
  properties: { maxLength: { type: 'number' }, unit: { type: 'string', enum: ['unicode-code-points', 'utf8-bytes'] },
    countWhitespace: { type: 'boolean' } },
}
export const FORMAT_VALIDATION_SCHEMA: JsonSchema = {
  type: 'object', additionalProperties: false,
  required: ['schema_version', 'status', 'constraints', 'passed', 'actual_length', 'output_sha256'],
  properties: {
    schema_version: { type: 'string', enum: ['output-length-check-v1'] },
    status: { type: 'string', enum: ['not-requested', 'not-evaluated', 'passed', 'failed'] },
    constraints: { oneOf: [OUTPUT_CONSTRAINTS_SCHEMA, { type: 'null' }] },
    passed: { oneOf: [{ type: 'boolean' }, { type: 'null' }] },
    actual_length: { oneOf: [{ type: 'number' }, { type: 'null' }] },
    output_sha256: { oneOf: [{ type: 'string' }, { type: 'null' }] },
  },
}

File: src/refractrouter/output_constraints.py
"""显式最终输出长度约束；计数、验收和原文保存互不混用。"""
import hashlib


def validate_output_constraints(raw):
    if not isinstance(raw, dict) or set(raw) != {'maxLength', 'unit', 'countWhitespace'}:
        raise ValueError('outputConstraints requires maxLength, unit and countWhitespace')
    if type(raw['maxLength']) is not int or not 1 <= raw['maxLength'] <= 1000000:
        raise ValueError('outputConstraints.maxLength must be an integer in 1..1000000')
    if raw['unit'] not in ('unicode-code-points', 'utf8-bytes'):
        raise ValueError('outputConstraints.unit must be unicode-code-points or utf8-bytes')
    if type(raw['countWhitespace']) is not bool:
        raise ValueError('outputConstraints.countWhitespace must be boolean')
    return dict(raw)


def check_output_constraints(constraints, answer=None):
    result = {'schema_version': 'output-length-check-v1',
              'status': 'not-requested' if constraints is None else 'not-evaluated',
              'constraints': constraints, 'passed': None, 'actual_length': None,
              'output_sha256': None}
    if constraints is None or answer is None:
        return result
    constraints = validate_output_constraints(constraints)
    measured = answer if constraints['countWhitespace'] else ''.join(c for c in answer if not c.isspace())
    actual = len(measured.encode('utf-8')) if constraints['unit'] == 'utf8-bytes' else len(measured)
    passed = actual <= constraints['maxLength']
    return {**result, 'status': 'passed' if passed else 'failed', 'passed': passed,
            'actual_length': actual, 'output_sha256': hashlib.sha256(answer.encode('utf-8')).hexdigest()}


def output_constraint_instruction(constraints):
    unit = 'Unicode 码点（不是 token 或视觉字形）' if constraints['unit'] == 'unicode-code-points' else 'UTF-8 字节'
    spaces = '计入所有空白' if constraints['countWhitespace'] else '计数前排除 Python str.isspace() 识别的空白'
    return (f"最终交付正文最多 {constraints['maxLength']} 个{unit}；{spaces}。"
            '数字、标点和 Markdown 标记均计入；只约束最终交付正文，不约束中间节点。')
